### ISSUE TITLE
System.TermInfoReader cannot handle new NCurses 6 TermInfo files

### ISSUE THREAD

**Issue by directhex** [mono/mono#6752]
## Steps to Reproduce

1. Install NCurses 6.1 (e.g. on Arch)
2. Try to do anything, e.g. make runtime

## Current Behavior

Crashes

### On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

Any
### Stacktrace

```Unhandled Exception:
System.TypeInitializationException: The type initializer for 'System.Console' threw an exception. ---> System.TypeInitializationException: The type initializer for 'System.ConsoleDriver' threw an exception. ---> System.Exception: Magic number is wrong: 542
  at System.TermInfoReader.ReadHeader (System.Byte[] buffer, System.Int32& position) [0x00028] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.TermInfoReader..ctor (System.String term, System.String filename) [0x0005f] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.TermInfoDriver..ctor (System.String term) [0x00055] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.ConsoleDriver.CreateTermInfoDriver (System.String term) [0x00000] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.ConsoleDriver..cctor () [0x0004d] in <10f1e38af51947f98c2267bd20fa8c60>:0 
   --- End of inner exception stack trace ---
  at System.Console.SetupStreams (System.Text.Encoding inputEncoding, System.Text.Encoding outputEncoding) [0x00007] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.Console..cctor () [0x0008e] in <10f1e38af51947f98c2267bd20fa8c60>:0 
   --- End of inner exception stack trace ---
  at Microsoft.CodeAnalysis.CommandLine.BuildClient.RunCompilation (System.Collections.Generic.IEnumerable`1[T] originalArguments, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter) [0x00005] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
  at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.Run (System.Collections.Generic.IEnumerable`1[T] arguments, Microsoft.CodeAnalysis.CommandLine.RequestLanguage language, Microsoft.CodeAnalysis.CommandLine.CompileFunc compileFunc, Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader analyzerAssemblyLoader) [0x00037] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.MainCore (System.String[] args) [0x00017] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.Main (System.String[] args) [0x00000] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeInitializationException: The type initializer for 'System.Console' threw an exception. ---> System.TypeInitializationException: The type initializer for 'System.ConsoleDriver' threw an exception. ---> System.Exception: Magic number is wrong: 542
  at System.TermInfoReader.ReadHeader (System.Byte[] buffer, System.Int32& position) [0x00028] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.TermInfoReader..ctor (System.String term, System.String filename) [0x0005f] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.TermInfoDriver..ctor (System.String term) [0x00055] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.ConsoleDriver.CreateTermInfoDriver (System.String term) [0x00000] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.ConsoleDriver..cctor () [0x0004d] in <10f1e38af51947f98c2267bd20fa8c60>:0 
   --- End of inner exception stack trace ---
  at System.Console.SetupStreams (System.Text.Encoding inputEncoding, System.Text.Encoding outputEncoding) [0x00007] in <10f1e38af51947f98c2267bd20fa8c60>:0 
  at System.Console..cctor () [0x0008e] in <10f1e38af51947f98c2267bd20fa8c60>:0 
   --- End of inner exception stack trace ---
  at Microsoft.CodeAnalysis.CommandLine.BuildClient.RunCompilation (System.Collections.Generic.IEnumerable`1[T] originalArguments, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter) [0x00005] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
  at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.Run (System.Collections.Generic.IEnumerable`1[T] arguments, Microsoft.CodeAnalysis.CommandLine.RequestLanguage language, Microsoft.CodeAnalysis.CommandLine.CompileFunc compileFunc, Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader analyzerAssemblyLoader) [0x00037] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.MainCore (System.String[] args) [0x00017] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.Main (System.String[] args) [0x00000] in <0d46d413edc9411fa4eb1c4cbce40409>:0 
```

Relevant, is `file`'s support for recognising the new files: https://github.com/file/file/commit/8f8d2c04892f58781f32907e4032dbac59df8625#diff-fe71ced20976bbadd001fc32fbea03a7

I'm not sure how much deeper the changes go

**Comment by directhex** [comment]
Looks like a big ol' new ABI in NCurses 6. https://www.gnu.org/software/ncurses/

**Comment by gfhcs** [comment]
Same problem on OpenSUSE Tumbleweed.

**Comment by gfhcs** [comment]
Workaround: On OpenSUSE Tumbleweed I was able to downgrade the packages "terminfo" and "terminfo-base" from 6.1 to 5.9 without breaking any (encoded) dependencies. Now terminal output with mono works again.

I just hope that I didn't break anything with this downgrade, 'cause I have no clue what terminfo is used for...

**Comment by migueldeicaza** [comment]
Slang added support for this recently: https://github.com/hankem/S-Lang/commit/6dd5ade9a97b52ace4ac033779a6d3c1c51db4d1

**Comment by lewurm** [comment]
fix is in progress. A simple workaround is to set `TERM=xterm` for example, that is, some terminfo file that is in the old file format.

**Comment by sbrl** [comment]
Any timeline on when this will be included in a release?

**Comment by lewurm** [comment]
@marek-safar should we backport this change to release branches? I'm not sure about it, given there is a fairly simple workaround for it.

**Comment by marek-safar** [comment]
@lewurm seems to be a common issue, could you backport it to 2018-02 ?

**Comment by cryptomilk** [comment]
Look at the stacktrace you printed (see the initial description). Do you think it is obvious for a user what the problem and workaround is?

I think the answer so no, so please backport ... :-)

**Comment by sbrl** [comment]
Yeah, it took me 12 hour of googling & at least a dozen web pages before I stumbled across this bug by chance.

**Comment by lewurm** [comment]
https://github.com/mono/mono/pull/7308 got merged, thus this fix will be included in 5.12

**Comment by sbrl** [comment]
@lewing I see. Quite the wait then! Thanks.

**Comment by ixvc** [comment]
Haha. Gotta use xterm till then ><

**Comment by alwayslivid** [comment]
I'm still dealing with the exact same issue. (in Arch Linux)

**Comment by directhex** [comment]
The fix for this was backported to Mono 5.10.x, and is in 5.10.1.45 as a minimum, which is in the mono-project.com `preview` and `stable` release branches (it is not yet included in a `vs` branch package)

**Comment by hussamT** [comment]
On Wed, 2018-04-18 at 21:03 -0700, Panos V. wrote:
> I'm still dealing with the exact same issue. (in Arch Linux)
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

If you can, rebuild the arch linux package after editing the PKGBUILD
and changing the _gitcommit variable to
b4e428d7c410dba6973d84ec13532debb5535d78

**Comment by stehgold** [comment]
Does anyone know why Fedora comes with 4.8. If this was fixes in 5.10 when will this arrive in Fedora, F30 ?

**Comment by HinTak** [comment]
I can think of one reason - mono 5 has many incompatible(?) changes / additions, including a new c# compiler contributed by Microsoft. In fact I am not upgrading until one (the ability to build ironpython correctly) is fixed.
But the ncurse issue in the subject is already backported to 4.8 (at least as distributed by fedora) and fixed.
    On Thursday, 12 July 2018, 20:00:50 GMT+8, Stephan Goldenberg <notifications@github.com> wrote:  
 
 
Does anyone know why Fedora comes with 4.8. If this was fixes in 5.10 when will this arrive in Fedora, F30 ?
